### PR TITLE
Perception: change darkscnn caffe model to tensorrt & add new obstacl…

### PR DESCRIPTION
**Pull request by xiaoMrzhang**
…e module

### CODE REVIEW

File: modules/perception/camera/lib/obstacle/detector/smoke/region_output.cc [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  for (auto &obj : *objects) {
+    float xmin = obj->camera_supplement.box.xmin;
+    float ymin = obj->camera_supplement.box.ymin;
+    float xmax = obj->camera_supplement.box.xmax;
+    float ymax = obj->camera_supplement.box.ymax;
+    int x = static_cast<int>(xmin * static_cast<float>(roi_w));
+    int w = static_cast<int>((xmax - xmin) * static_cast<float>(roi_w));
+    int y = static_cast<int>(ymin * static_cast<float>(roi_h)) + offset_y;
+    int h = static_cast<int>((ymax - ymin) * static_cast<float>(roi_h));
+    base::RectF rect_det(static_cast<float>(x), static_cast<float>(y),
```
Comment (storypku) [inline modules/perception/camera/lib/obstacle/detector/smoke/region_output.cc:60]: Q: why x, y, w, h first cast into int then into float?

File: modules/perception/camera/lib/obstacle/detector/smoke/smoke_obstacle_detector.cc [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+  }
+  InitSmokeBlob(smoke_param_.net_param());
+  if (!InitFeatureExtractor(model_root)) {
+    return false;
+  }
+  return true;
+}
+
+bool SmokeObstacleDetector::InitFeatureExtractor(const std::string &root_dir) {
+  FeatureExtractorInitOptions feat_options;
```
Comment (storypku) [inline modules/perception/camera/lib/obstacle/detector/smoke/smoke_obstacle_detector.cc:195]: just name it feature_options or options. `feat` is not an abbr of feature.

File: modules/perception/camera/lib/obstacle/detector/smoke/object_maintainer.cc [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+bool SmokeObjectMaintainer::Add(int idx, base::ObjectPtr obj) {
+  auto obj_it = assigned_index_.find(idx);
+  if (obj_it == assigned_index_.end()) {
+    assigned_index_[idx] = obj;
+    return true;
+  }
+
+  auto prev_obj = obj_it->second;
+  const auto &&curr_type = static_cast<int>(obj->sub_type);
```
Comment (changsh726) [inline modules/perception/camera/lib/obstacle/detector/smoke/object_maintainer.cc:33]: why use rvalue reference here

File: modules/perception/production/data/perception/camera/models/traffic_light_recognition/horizontal/deploy.prototxt [hunk 4]
```diff
@@ -105,7 +105,7 @@ layer {
     stride_h: 2
     pad_w: 1
     pad_h: 1
-    round_mode: 1
+    cmp_out_shape_floor_as_conv: true
```
Comment (jeroldchen) [inline modules/perception/production/data/perception/camera/models/traffic_light_recognition/horizontal/deploy.prototxt:108]: 1. I noticed that you haven't modify the corresponding `addPoolingLayer` function in `rt_net.cc`. Only if you specify this param by TensorRT IPoolingLayer's setPaddingMode(nvinfer1::PaddingMode::kCAFFE_ROUND_DOWN), can it be truly loaded. [https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_pooling_layer.html#ad0fabf28e72fb4c0f5ec5692444a79bf](url)
2. Therefore, there is no need to change the prototxt. Just keep all the round_mode usage unchanged. Instead, you should modify rt.proto, as Caffe already updated round_mode param in the proto: [https://github.com/BVLC/caffe/pull/6282/files](url)
3. The meaning of round_mode and TensorRT's corresponding param is explained here: [https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/namespacenvinfer1.html#a72f43f32e90e4ac5548f8c9ae007584c](url)

Comment (jeroldchen) [reply modules/perception/production/data/perception/camera/models/traffic_light_recognition/horizontal/deploy.prototxt:108]: And `cmp_out_shape_floor_as_conv` is only used to change the name of layer, it doesn't have any function.

**Review by guoweiwan — APPROVED** [review]

**Pull request merged**